File: Tree/572.Subtree-of-Another-Tree/572.Subtree-of-Another-Tree.ts
function isSubtree(root: TreeNode | null, subRoot: TreeNode | null): boolean {
  if (root === null || subRoot === null) {
    return root === subRoot;
  }

  if (
    isSameTree(root, subRoot) ||
    isSubtree(root.left, subRoot) ||
    isSubtree(root.right, subRoot)
  ) {
    return true;
  } else {
    return false;
  }
}
function isSameTree(root: TreeNode | null, subRoot: TreeNode | null): boolean {
  if (root === null || subRoot === null) {
    return root === subRoot;
  }

  if (
    root.val === subRoot.val &&
    isSameTree(root.left, subRoot.left) &&
    isSameTree(root.right, subRoot.right)
  ) {
    return true;
  } else {
    return false;
  }
}

File: Tree/572.Subtree-of-Another-Tree/572.Subtree-of-Another-Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root or not subRoot:
            return root == subRoot
        elif self.isSameTree(root, subRoot):
            return True
        else:
            return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
            
    def isSameTree(self, tree1, tree2):
        if not tree1 or not tree2:
            return tree1 == tree2
            
        return tree1.val == tree2.val and self.isSameTree(tree1.left, tree2.left) and self.isSameTree(tree1.right, tree2.right)
